src/conversion/onnx_to_krnl/nn/pooling.cpp

```diff
@@ -222,45 +222,107 @@ struct ONNXMaxPoolSingleOutOpLowering : public ConversionPattern {
           }
+
+          // sX * rX + kX
           spatialIndex = rewriter.create<AddIOp>(
               loc, spatialIndex, innerLoops.getInductionVar(i - batchRank));
-          // If ceil mode is enabled, then the calculated access index may
-          // exceed its dimension. In such a case, we will use the maximum
-          // index, which causes multiple visits to the element of the
-          // maximum index.
-          // TODO: Avoid multiple visits.
-          if (ceilMode) {
-            Value inputIndex;
+
+          // Dilate the kernel index only if the dilation value is not one (not
+          // default). If the kernel dimension has a real center, keep the index
+          // at the center unchanged/undilated.
+          if (dilations[i - batchRank] > 1) {
+            bool hasCenter = (kernelShape[i - batchRank] % 2) == 1;
+            int64_t centerIndex = (hasCenter)
+                                      ? ((kernelShape[i - batchRank] - 1) / 2)
+                                      : (kernelShape[i - batchRank] / 2);
+            auto centerKernelIndex = emitConstantOp(
+                rewriter, loc, rewriter.getIndexType(), centerIndex);
+            Value kernelIndex = innerLoops.getInductionVar(i - batchRank);
+
+            Value lessThanCenter, greaterThanCenter;
+            if (hasCenter) {
+              lessThanCenter = rewriter.create<CmpIOp>(
+                  loc, CmpIPredicate::slt, kernelIndex, centerKernelIndex);
+              greaterThanCenter = rewriter.create<CmpIOp>(
+                  loc, CmpIPredicate::sgt, kernelIndex, centerKernelIndex);
+            } else {
+              lessThanCenter = rewriter.create<CmpIOp>(
+                  loc, CmpIPredicate::slt, kernelIndex, centerKernelIndex);
+              greaterThanCenter = rewriter.create<CmpIOp>(
+                  loc, CmpIPredicate::sge, kernelIndex, centerKernelIndex);
+            }
+
+            auto dilationIndex = emitConstantOp(rewriter, loc,
+                rewriter.getIndexType(), dilations[i - batchRank] - 1);
+            // Dilate forward.
+            auto spatialAddDilationIndex =
+                rewriter.create<AddIOp>(loc, spatialIndex, dilationIndex);
+            // Dilate backward.
+            auto spatialMinusDilationIndex =
+                rewriter.create<SubIOp>(loc, spatialIndex, dilationIndex);
+            // Switch between forward and backward mode depending on
+            // relationship between the spatial index and the center index.
+            spatialIndex = rewriter.create<SelectOp>(
+                loc, lessThanCenter, spatialMinusDilationIndex, spatialIndex);
+            spatialIndex = rewriter.create<SelectOp>(
+                loc, greaterThanCenter, spatialAddDilationIndex, spatialIndex);
+          }
+
+          // Check whether the spatial index goes out of the input index space
+          // or not.
+          if (dilations[i - batchRank] > 1 or ceilMode) {
```

Comment:
Can you give an example of the code that is generated when out of bound, in a comment?
If out of bound, do you generate an affine-if?

Reply:
Now we only have out-of-bound when the spatial index is greater than the upper bound of the input dimension. This may happen when we convolve, say, bottom rows and right columns of a 2D array.